[babel-preset-expo] Add an option `lazyImports` for lazy-initialize/inline-require

hesyifei opened the pull request:
# Why

Making `import` statements to be lazily evaluated (i.e., inline required) when their imported bindings are used for the first time. This can improve the initial load time of the app because evaluating dependencies up front is sometimes entirely unnecessary. This is especially the case when implementing a library module. (Also refer to https://github.com/facebook/metro/commit/23e3503dde5f914f3e642ef214f508d0a699851d.)

# How

`babel-preset-expo` is using [metro-react-native-babel-preset](https://github.com/facebook/metro/tree/master/packages/metro-react-native-babel-preset) which has an option `lazyImportExportTransform` that is then passed into [@babel/plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy)'s option `lazy`.

The value of `lazyImports` in `babel-preset-expo` has similar meaning to the `lazy` option in `@babel/plugin-transform-modules-commonjs`.

- `null` - [metro-react-native-babel-preset](https://github.com/facebook/metro/tree/master/packages/metro-react-native-babel-preset) will handle it.

- `false` - No lazy initialization of any imported module.

- `true` - Lazy-init all imported modules except local imports (e.g., `./foo`), certain Expo packages that have side effects, and the two cases mentioned [here](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy).

- `Array<string>` - [babel-plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) will handle it.

- `(string) => boolean` - [babel-plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) will handle it.

---

When doing inline require, we have to consider the effect of the code that is outside a class or a function. For modules with these side effects, we have to require them at the beginning of the file and cannot inline require them. Some (but not all - see #4659) such modules in Expo are marked with `.fx.js` or `.fx.ts` suffix. I wrote a small [script](https://gist.github.com/hesyifei/eb293983260ecc9e77de48133181f2e9) to look at the code in the `src/` folder in each package (in `packages/` folder) and find those with "CallExpression" or "IfStatement" at the top level (i.e., not inside any class or function). These modules are stored in `lazy-imports-blacklist.js`. While I am not confident that this list is entirely correct, I have tested `apps/native-component-list/` and it seems to be working correctly.

# Test Plan

I have added tests for the `lazyImports` option and the generated code is as intended. To better test change, we should try to add `lazyImports: true` to apps such as `home/`, `apps/storybook/`, `apps/standalone-ncl/`, and `apps/storybook/`.

hesyifei committed:
Update "metro-react-native-babel-preset" to 0.54.1 to support `lazyImportExportTransform`

Ref: https://github.com/facebook/metro/commit/23e3503dde5f914f3e642ef214f508d0a699851d

hesyifei committed:
[babel-preset-expo] Add an option `lazy` for lazy-initialize/inline-require

`lazy` is default to be `null`.

If `lazy` is set to `true`, then all packages except local imports and certain expo packages in the blacklist (due to their side effects) will be lazy-initialized.
If `lazy` is set to `null`, `metro-react-native-babel-preset` will handle it.
If `lazy` is set to any other value, that value be passed directly to the `lazy` option of `babel-plugin-transform-modules-commonjs`. As such, it can be `false`, `Array<string>`, or `(string) => boolean`.

Ref: https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy and https://github.com/facebook/metro/commit/23e3503dde5f914f3e642ef214f508d0a699851d

hesyifei committed:
[babel-preset-expo] Add documentation for `lazy`

hesyifei committed:
[babel-preset-expo] Clarify `lazy: true` in the documentation

hesyifei committed:
[babel-preset-expo] Add tests for `lazy` option

hesyifei committed:
[home] Enable `lazy` for `babel-preset-expo`

ide commented on packages/babel-preset-expo/__tests__/index-test.js:
```diff
@@ -72,3 +72,48 @@ import 'react-native-vector-icons';
   expect(code).toMatchSnapshot();
 });
+
+it(`uses lazy - default (no option) matches option "null"`, () => {
```
Wrap these new tests in:
```
describe('"lazy" option', () => {
  it (`defaults to null`, () => ... );
});
```

Jest will print out `"lazy" options defaults to null` in its test output for you.

ide commented on packages/babel-preset-expo/__tests__/index-test.js:
```diff
@@ -72,3 +72,48 @@ import 'react-native-vector-icons';
+  expect(codeDefault).toEqual(codeNull);
+});
+
+it(`uses lazy - different values`, () => {
```
For a test like this where the code is the same and the inputs are just slightly different, you can use:
```
describe.each(
   [[null],
    [false],
    [true],
    [['inline-comp', './inline-func', '../inline-func-with-side-effects.fx.ts']],
    [name => !(name.endsWith('.fx') || name.endsWith('.fx.js') || name.endsWith('.fx.ts'))],
  ])(`accepts %p`, lazyOption => {
  ... test
});
```
```

ide commented on packages/babel-preset-expo/index.js:
```diff
@@ -1,16 +1,38 @@
+              ? require_name => {
+                  // Do not lazy-initialize packages that are local imports (similar to `lazy: true` behavior)
+                  // or are in the blacklist.
+                  return !(require_name.includes('./') || lazyImportsBlacklist.has(require_name));
```
We would like to lazily import an app's own modules (ex: App.js should lazily import `./screens/TermsOfServiceScreen`) too -- how could we enable that?

hesyifei replied:
On `@babel/plugin-transform-modules-commonjs`'s [docs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) for the `lazy` option, it says
> Local paths are much more likely to have circular dependencies, which may break if loaded lazily, so they are not lazy by default, whereas dependencies between independent modules are rarely cyclical.

Also, I found that lazily importing local modules broke the Back navigation button in `apps/native-component-list/`. So I think it would be better to not lazily load local modules by default. If the developer wants to lazily import an app's own modules, they could use `(string) => boolean` to handle it themselves.

(Though if we are doing that, we should probably find a way to provide `lazyImportsBlacklist` to the developer.)

hesyifei replied:
I added a note here: https://github.com/expo/expo/pull/4685/commits/8cf5a62c416e90ff78b3bdf12e3f32da8b2b531e#diff-cc0648e61fc1cd454f9938303cc8d00d

ide commented on packages/babel-preset-expo/lazy-imports-blacklist.js:
```diff
@@ -0,0 +1,43 @@
+ */
+'use strict';
+
+module.exports = new Set([
```
This is a large set of packages -- let's check each one to see if the side effects actually exist and are important.

hesyifei replied:
I've checked each package and removed the false positives.

hesyifei commented on packages/babel-preset-expo/index.js:
```diff
@@ -1,16 +1,38 @@
+    presets: [
+      [
+        // TODO: use the actual plugin - but somehow `module:metro-react-native-babel-preset` is not up-to-date yet (need a version after this commit: https://github.com/facebook/metro/commit/23e3503dde5f914f3e642ef214f508d0a699851d)
+        require('./node_modules/metro-react-native-babel-preset'),
```
If I just use `['module:metro-react-native-babel-preset']`, it still uses v0.51.1 which does not support custom `lazyImportExportTransform` (which is added in v0.53.1). I have update its version in all related `package.json` (https://github.com/expo/expo/pull/4685/commits/0f31f4c26ef69ceafbd4c05742dcaec67baeaa68) but it seems that some other dependencies are depend on v0.51.1, so `['module:metro-react-native-babel-preset']` uses v0.51.1. Maybe we need to also update those dependencies?

hesyifei committed:
[babel-preset-expo] Fix variable names

hesyifei committed:
[babel-preset-expo] Update test format

hesyifei committed:
Update CHANGELOG.md

hesyifei committed:
[babel-preset-expo] Remove packages that don't have side-effects from lazy-imports-blacklist.js

Previously the script (https://gist.github.com/hesyifei/eb293983260ecc9e77de48133181f2e9) forgot to exclude "__test__" files

hesyifei committed:
[babel-preset-expo] Set version to "5.2.1"

hesyifei committed:
[babel-preset-expo] Update comment about `require('metro-react-native-babel-preset')`

hesyifei committed:
[babel-preset-expo] Add notes for how to access `lazyImportsBlacklist`

Ref: https://github.com/expo/expo/pull/4685#discussion_r296462657

hesyifei committed:
[babel-preset-expo] Minor formatting change

hesyifei committed:
[babel-preset-expo] Remove packages that don't have side-effects from lazy-imports-blacklist.js

Previously the script (https://gist.github.com/hesyifei/eb293983260ecc9e77de48133181f2e9) did not exclude "__export" functions (which are not side-effects)

hesyifei committed:
[babel-preset-expo] Remove packages that don't have side-effects from lazy-imports-blacklist.js

Previously the script (https://gist.github.com/hesyifei/eb293983260ecc9e77de48133181f2e9) did not exclude `export enum` and `export type` related statements (which are not side-effects)

hesyifei committed:
[babel-preset-expo] Remove packages that don't have side-effects from lazy-imports-blacklist.js

Previously the script (https://gist.github.com/hesyifei/eb293983260ecc9e77de48133181f2e9) did not exclude `Object.assign` statements (which are not side-effects)

hesyifei committed:
[babel-preset-expo] Remove packages that don't have side-effects from lazy-imports-blacklist.js

Checked manually and removed all unnecessary packages in the blacklist

Related files:
```
expo/packages/expo-asset/src/Asset.ts
expo/packages/expo-background-fetch/src/BackgroundFetch.ts
expo/packages/expo-constants/src/Constants.ts
expo/packages/expo-file-system/src/ExponentFileSystem.ts
expo/packages/expo-file-system/src/FileSystem.ts
expo/packages/expo-file-system/src/index.ts
expo/packages/expo-location/src/Location.ts
expo/packages/expo-sqlite/src/polyfillNextTick.ts
expo/packages/expo-sqlite/src/SQLite.ts
expo/packages/expo-task-manager/src/TaskManager.ts
expo/packages/jest-expo/src/preset/setup.js
expo/packages/expo-font/src/Font.ts
expo/packages/expo/src/deprecated.ts
expo/packages/expo/src/environment/logging.fx.ts
expo/packages/expo/src/environment/validate.fx.ts
expo/packages/expo/src/Expo.fx.ts
expo/packages/expo/src/Expo.ts
expo/packages/expo/src/globals.ts
expo/packages/expo/src/Linking/Linking.ts
expo/packages/expo/src/Notifications/Notifications.ts
expo/packages/expo/src/timer/polyfillNextTick.ts
```

hesyifei committed:
[babel-preset-expo] Format blacklist

hesyifei committed:
Merge branch 'master' into @hesyifei/inline-require

esamelson reviewed (approved):
i think this looks good, let's also open issues (if we don't already have them) to remove the side effects from the modules in your blacklist, and move them into .fx.ts files.

esamelson commented:
actually, we also need to make sure this doesn't break anything on web

ide commented on packages/babel-preset-expo/index.js:
```diff
@@ -1,16 +1,42 @@
+      [
+        // TODO: use `module:metro-react-native-babel-preset` after this package is updated to at least v0.53.1 in the root `node_modules`.
+        // Ref: https://github.com/expo/expo/pull/4685#discussion_r296462573
+        require('metro-react-native-babel-preset'),
```
Maybe let's always do this, even if we upgrade the preset elsewhere. It's not good if the presets can change unintentionally.

hesyifei committed:
Merge branch 'master' into @hesyifei/inline-require

hesyifei committed:
Minor change in formatting

hesyifei committed:
[babel-preset-expo] Add notes for `require('metro-react-native-babel-preset')`

hesyifei committed:
[babel-preset-expo] Rename `lazy` option to `lazyImports`

hesyifei committed:
[babel-preset-expo] Clean up for renaming `lazy` option to `lazyImports`

hesyifei committed:
[ncl] Add lazyImports

hesyifei committed:
Update packages/babel-preset-expo/README.md

Co-Authored-By: James Ide <ide@users.noreply.github.com>

hesyifei committed:
[babel-preset-expo] Minor change in test

Co-Authored-By: James Ide <ide@users.noreply.github.com>

ide commented on packages/babel-preset-expo/README.md:
```diff
@@ -5,6 +5,37 @@ at https://github.com/expo/expo. Thanks!
+### [`lazyImports`](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy)
+
+Changes Babel's compiled `import` statements to be lazily evaluated when their imported bindings are used for the first time.
+
```
```suggestion

*Note:* this option has an effect only when the `disableImportExportTransform` option is set to `false`. On Android and iOS, `disableImportExportTransform` defaults to `false`, and on web it defaults to `true` to allow for tree shaking.

```

ide reviewed (approved):
Add a note about how this option works on web (see inline comment). Looks good otherwise.

hesyifei committed:
Update packages/babel-preset-expo/README.md

Co-Authored-By: James Ide <ide@users.noreply.github.com>
